mathjax3-ts/input/tex/NodeFactory.ts

```diff
@@ -0,0 +1,188 @@
+        if (child.isInferred) {
+          let mrow = factory.mmlFactory.create('mrow', {}, NodeUtil.getChildren(child));
+          NodeUtil.copyAttributes(child, mrow);
+          cleanChildren.push(mrow);
+        } else {
+          cleanChildren.push(child);
+        }
+      }
+      node.setChildren(cleanChildren);
+    }
+    if (text) {
+      node.appendChild(text);
+    }
+    NodeUtil.setProperties(node, def);
+    return node;
+  };
+
+
+  /**
+   * Default token generation function.
+   * @param {NodeFactory} factory The current node factory.
+   * @param {string} kind The type of node to create.
+   * @param {any} def Its properties.
+   * @param {string} text Text of the token.
+   * @return {MmlNode} The newly created token node.
+   */
+  public static createToken(factory: NodeFactory, kind: string,
+                            def: any = {}, text: string = ''): MmlNode  {
+    const textNode = factory.create('text', text);
+    return factory.create('node', kind, [], def, textNode);
+  }
+
+
+  /**
+   * Default text node generation function.
+   * @param {NodeFactory} factory The current node factory.
+   * @param {string} text The text for the new node.
+   * @return {TextNode} The newly created text node.
+   */
+  public static createText(factory: NodeFactory, text: string): TextNode  {
+    if (text == null) {
+      return null;
+    }
+    return (factory.mmlFactory.create('text') as TextNode).setText(text);
+  };
+
+
+  /**
+   * Default error node generation function.
+   * @param {NodeFactory} factory The current node factory.
+   * @param {string} message The error message.
+   * @return {MmlNode} The newly created error node.
+   */
+  public static createError(factory: NodeFactory, message: string): MmlNode  {
+    let text = factory.create('text', message);
+    let mtext = factory.create('node', 'mtext', [], {}, text);
+    let error = factory.create('node', 'merror', [mtext]);
+    return error;
+  };
+
+
+  /**
+   * Adds a method to the factory.
+   * @param {string} kind The type of node the method creates.
+   * @param {NodeFactoryMethod} func The node creator.
+   */
+  public set(kind: string, func: NodeFactoryMethod) {
+    this.factory[kind] = func;
+  }
+
+
+  /**
+   * Adds a set of node creators to the factory.
+   * @param {Object.<NodeFactoryMethod>} maps The set of functions.
+   */
+  public setCreators(maps: {[kind: string]: NodeFactoryMethod}) {
+    for (let kind in maps) {
+      this.set(kind, maps[kind]);
+    }
+  }
+
+
+  /**
+   * Creates a node for the internal data structure from the factory.
+   * @param {string} kind The type of node to be created.
+   * @param {any[]} ...rest The arguments for the node.
+   * @return {MmlNode} The created node.
+   */
+  public create(kind: string, ...rest: any[]): MmlNode {
+    const func = this.factory[kind] || this.factory['node'];
+    const node = func(this, rest[0], ...rest.slice(1));
+    this.configuration.addNode(rest[0], node);
```

Comment:
Here's my reasoning:
As we are only saving pointers anyway, it won't cost much memory. And we are pushing into an O(1) data structure. On the other hand, testing which node to save or not, might cost a fraction of time (string comparison). 
Furthermore, if an extension package wants to implement a postfilter on a particular node class, it can just use them without having to jump through any hoops, like adding a new `NodeFactory` method.
In fact, I have started playing with some for a11y information already.

Reply:
OK, I thought it might be that you were doing it for possible future uses.